### PR TITLE
refactor: allow the content of `InstantiatedChunk` to be binary data

### PR DESCRIPTION
**Pull request by hyf0**

### CODE REVIEW

**Comment by hyf0** [comment]
* **#2562** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2562?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2561** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2561?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2560** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2560?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2559** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2559?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2558** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2558?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2557** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2557?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2539** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/rolldown/src/utils/render_chunks.rs [hunk 1]
```diff
@@ -18,12 +18,12 @@ pub async fn render_chunks<'a>(
     if let InstantiationKind::Ecma(ecma_meta) = &asset.kind {
       let render_chunk_ret = plugin_driver
         .render_chunk(HookRenderChunkArgs {
-          code: asset.content.clone(),
+          code: asset.content.clone().try_into_string()?,
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/utils/render_chunks.rs:21]: Why do we need to call the hook if it content is none string? Did rollup do that?

File: crates/rolldown/src/asset/asset_generator.rs [hunk 2]
```diff
@@ -32,7 +32,7 @@ impl Generator for AssetGenerator {
       instantiated_chunks.push(InstantiatedChunk {
         origin_chunk: ctx.chunk_idx,
         // TODO: should support buffer here
-        content: String::from_utf8(asset_view.source.to_vec()).unwrap(),
+        content: StrOrBytes::Bytes(asset_view.source.to_vec()),
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/asset/asset_generator.rs:35]: I think you could remove the todo now

File: crates/rolldown/src/utils/render_chunks.rs (hunk 1 above)
Comment (hyf0) [reply crates/rolldown/src/utils/render_chunks.rs:21]: image
For ecam chunk, the content will always be string. `Bytes` branch should be unreachable.

File: crates/rolldown/src/asset/asset_generator.rs (hunk 2 above)
Comment (hyf0) [reply crates/rolldown/src/asset/asset_generator.rs:35]: It's done by later PRs.

File: crates/rolldown/src/utils/render_chunks.rs (hunk 1 above)
Comment (hyf0) [reply crates/rolldown/src/utils/render_chunks.rs:21]: I used a more proper method in https://github.com/rolldown/rolldown/pull/2559.

**Review by IWANABETHATGUY — APPROVED** [review]

**Pull request merged**